index/scorch/snapshot_rollback.go

```diff
@@ -105,69 +116,85 @@ func (s *Scorch) RollbackPoints() ([]*RollbackPoint, error) {
 	return rollbackPoints, nil
 }
 
-// Rollback atomically and durably (if unsafeBatch is unset) brings
-// the store back to the point in time as represented by the
-// RollbackPoint. Rollback() should only be passed a RollbackPoint
-// that came from the same store using the RollbackPoints() API.
-func (s *Scorch) Rollback(to *RollbackPoint) error {
-	if to == nil {
-		return fmt.Errorf("Rollback: RollbackPoint is nil")
+// Rollback atomically and durably brings the store back to the point
+// in time as represented by the RollbackPoint.
+// Rollback() should only be passed a RollbackPoint that came from the
```

Comment:
indeed, thats a wrinkle there... lemme add a validation there..

Reply:
Given the validations in place to guard against a possible index damage from an invalid rollback point, I don't see/foresee any problems with this destructive approach to bolt store.. But the pros are like 
-lean boltdb/store contents to manage
-faster disk clean ups on next index open as there is lesser epoch on bolt

Same time, I am unable to understand the benefits out of an extra epoch getting stacked to front approach too..

Reply:
1.  Is a lean boltdb really a concern? Once you open it and things start running, its going to return the NumSnapshotsToKeep.  And I've never seen a root.bolt ever get very large.  Does it in the use cases you support?

2. To me the only reason there is faster disk cleanup is that you reduce the root.bolt to a single snapshot (going below the users configuration).

3.  You had also mentioned some recursive rollback case, but it seemed to me you always process the RollbackPoints linearly looking for what you want, so I didn't see possibility of loop or recursion.

4.  To me, the best argument for your approach is that it is simple, and we got this wrong the first time.  I still like the idea of introducing a new snapshot that matches the rollback point, because then we go over the desired NumSnapshotsToKeep, instead of below, and once you open it, it returns to normal.  But I won't hold this up, and will let others give their input.

Reply:
> Is a lean boltdb really a concern? Once you open it and things start running, its going to return the NumSnapshotsToKeep. And I've never seen a root.bolt ever get very large. Does it in the use cases you support?

Less is better is the idea there.
We have seen boltdb containing thousands of items.. Its just about storing only the bare needed things in bolt at least whenever we get a clean up opportunity in form of rollback. 


> To me the only reason there is faster disk cleanup is that you reduce the root.bolt to a single snapshot (going below the users configuration).

We don't always go below the user configured numSnapshotsToKeep or to a single snapshot. We keep deleting epochs until we find the asked epoch. Going below user configuration happens if the asked epoch was the last remaining one in the bolt or the asked epoch was one among the retained numSnapshotsToKeep epochs.


> You had also mentioned some recursive rollback case, but it seemed to me you always process the RollbackPoints linearly looking for what you want, so I didn't see possibility of loop or recursion.

Repetitive rollback requests from KV was referred as the recursive possibility there. We should refrain from adding data into the system in such weird system behaviour.

Destructive approach fits the easy mental model of each epoch in bolt stores a linear progressive state of the index and the insert approach break it.   ;)

If the main concern is only about NumSnapshotsToKeep, 
then it isn't a significant worry to me as long as there is no math/logical relation exposed on the precise rollback-ability gained out of a chosen number of NumSnapshotsToKeep. 
Now there is an argument of what if once we fix it?   

But then we go below the NumSnapshotsToKeep only in two corner cases and that too under a successful rollback. So the rollback purpose is served. And then index open would start of with a clean disk of necessary files.

Reply:
One other thought popping to mind is - in the insert epoch approach ,
aren't we spoiling the quality/effectiveness of the retained  epochs for NumSnapshotsToKeep ?

As after a successful rollback, index starts from the rollbacked point and the epochs behind the newly introduced one aren't really helping/contributing to any of the next imminent rollback request to a point since the last successful rollback point?  

Epochs behind the recent rollback introduced one are going to be useful only for those cases - when the next upcoming rollback request is asking to rollback to even an older rollback point than the last rollback point, which is less often than rollback to a recent point since the last successful rollback point.. ? (debatable)